src/vs/code/electron-main/app.ts

```diff
@@ -846,6 +851,68 @@ export class CodeApplication extends Disposable {
 	}
 }
 
+class WebviewProtocolProvider extends Disposable {
+
+	private readonly webviewMetadata = new Map<string, {
+		readonly extensionLocation: URI | undefined;
+		readonly localResourceRoots: URI[];
+	}>();
+
+	constructor(
+		@IFileService private readonly fileService: IFileService,
+	) {
+		super();
+
+		ipc.on('vscode:registerWebview', (event: IpcMainEvent, id: string, data: any) => {
+			if (!this.webviewMetadata.has(id)) {
+				const ses = session.fromPartition(id);
+				this.register(ses);
+			}
+
+			this.webviewMetadata.set(id, {
+				extensionLocation: data.extensionLocation ? URI.from(data.extensionLocation) : undefined,
+				localResourceRoots: Array.isArray(data.localResourceRoots) ? data.localResourceRoots.map((x: UriComponents) => URI.from(x)) : []
+			});
+
+			event.sender.send(`vscode:didRegisterWebview-${id}`);
+		});
+
+		ipc.on('vscode:unregisterWebview', (_event: IpcMainEvent, id: string) => {
+			this.webviewMetadata.delete(id);
+		});
+	}
+
+	private register(session: Session) {
+		session.protocol.registerBufferProtocol(Schemas.vscodeWebviewResource, async (request, callback: any) => {
```

Comment:
@deepak1556 Each webview currently gets its own electron session since we give them [unique partitions](https://www.electronjs.org/docs/api/webview-tag#partition). Do you know if we would open up any security problems by having all webviews share the same session instead?

Reply:
As long as the webviews are origin isolated then there is no security concern with sharing the same session, they get the same level of security as the web defines.

You can think of session as an equivalent to a profile in chrome, a user of a profile can load websites from different origins and they all share the same data directory in the disk but at the implementation level all the storage access apis or network apis are guarded behind origin scopes. So a session or a profile is not a security feature, its more of an isolation of persisted data.

Reply:
Thanks. We load the webviews using a data uri so I think we are safe to stop using profiles. This simplifies this code